system/nxcamera/nxcamera.c

```diff
@@ -0,0 +1,881 @@
+
+  /* Save the path of the preferred device */
+
+  if (!(caps.capabilities & V4L2_CAP_VIDEO_CAPTURE))
+    {
+      return -ENODEV;
+    }
+
+  strlcpy(pcam->capturedev, device, sizeof(pcam->capturedev));
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nxcamera_setfb
+ *
+ *   nxcamera_setfb() sets the output framebuffer device to use with the
+ *   provided nxcamera context.
+ *
+ ****************************************************************************/
+
+int nxcamera_setfb(FAR struct nxcamera_s *pcam, FAR const char *device)
+{
+  int                    temp_fd;
+  struct fb_videoinfo_s  vinfo;
+
+  DEBUGASSERT(pcam != NULL);
+  DEBUGASSERT(device != NULL);
+
+  /* Try to open the device */
+
+  temp_fd = open(device, O_RDWR);
+  if (temp_fd == -1)
+    {
+      /* Error opening the device */
+
+      return -errno;
+    }
+
+  /* Validate it's a fb device by issuing an FBIOGET_VIDEOINFO ioctl */
+
+  if (ioctl(temp_fd, FBIOGET_VIDEOINFO, (uintptr_t)&vinfo) != OK)
+    {
+      /* Not an Video device! */
+
+      close(temp_fd);
+      return -errno;
+    }
+
+  /* Close the file */
+
+  close(temp_fd);
+
+  if (vinfo.nplanes == 0)
+    {
+      return -ENODEV;
+    }
+
+  /* Save the path of the framebuffer device */
+
+  strlcpy(pcam->displaydev, device, sizeof(pcam->displaydev));
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nxcamera_setfile
+ *
+ *   nxcamera_setfile() sets the output file path to use with the provided
+ *   nxcamera context.
+ *
+ ****************************************************************************/
+
+int nxcamera_setfile(FAR struct nxcamera_s *pcam, FAR const char *pfile,
+                     bool isimage)
+{
+  int temp_fd;
+
+  DEBUGASSERT(pcam != NULL);
+  DEBUGASSERT(pfile != NULL);
+
+  /* Try to open the file */
+
+  temp_fd = open(pfile, O_CREAT | O_RDWR);
+  if (temp_fd == -1)
+    {
+      /* Error opening the file */
+
+      return -errno;
+    }
+
+  /* Close the file */
+
+  close(temp_fd);
+
+  /* Save the path of the output file */
+
+  if (isimage)
+    {
+      strlcpy(pcam->oimagepath, pfile, sizeof(pcam->oimagepath));
+    }
+  else
+    {
+      strlcpy(pcam->ovideopath, pfile, sizeof(pcam->ovideopath));
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: nxcamera_stop
+ *
+ *   nxcamera_stop() stops the current playback to loop and closes the
+ *   file and the associated device.
+ *
+ * Input:
+ *   pcam    Pointer to the initialized looper context
+ *
+ ****************************************************************************/
+
+int nxcamera_stop(FAR struct nxcamera_s *pcam)
+{
+  struct video_msg_s term_msg;
+  FAR void           *value;
+
+  DEBUGASSERT(pcam != NULL);
+
+  /* Validate we are not in IDLE state */
+
+  sem_wait(&pcam->sem);                      /* Get the semaphore */
```

Comment:
seems that `sem` is used in mutex style. Maybe pthread mutex can be used instead?